File: 818-race-car/818-race-car.go
func racecar(target int) int {
    type pair struct {pos, vel int}
    var queue []pair
    var (
        currNode pair
        srcNode pair
    )
    srcNode = pair{0,1}
    queue = append(queue, srcNode)
    
    dist := 0 
    for len(queue) != 0 {
        //n := len(queue)
        for n := len(queue); n != 0; n-- {
            currNode, queue = queue[0], queue[1:]   
            if currNode.pos == target {
                return dist
            }
            pos, vel := currNode.pos, currNode.vel
            queue = append(queue, pair{pos+vel, vel*2})
            if pos + vel > target && vel > 0 {
                queue = append(queue, pair{pos, -1})
            } else if pos + vel < target && vel < 0 {
                queue = append(queue, pair{pos, 1})
            }
        }
        dist++
    }
    
    return -1
}

File: 818-race-car/818-race-car.cpp
class Solution {
public:
    
    int racecar(long int target) {
        queue<pair<int, int>> q;
        q.push({0, 1});
        int depth = 0;
        
        while (true) {
            //vector<pair<long int, long int>> next_q;
            int n = q.size(); 
            for (int idx = 0; idx < n; idx++) {
                pair<int, int> node = q.front();
                q.pop();
                int pos = node.first;
                long int speed = node.second;
                if (pos == target) {
                    return depth;
                }
                if (pos > 2 * target) continue;
                
                q.push({ pos + speed, 2 * speed });
                if (pos + speed > target && speed > 0) {
                    q.push({ pos, -1 });
                } else if (pos + speed < target && speed < 0) {
                    q.push({ pos, 1 });
                }
            }
            ++depth;
        }
        return -1;
    }
};